pkg/ingester/ingester.go

```diff
@@ -769,22 +778,35 @@ func (i *Ingester) PushWithCleanup(ctx context.Context, pushReq *push.Request) (
 
 	if len(req.EphemeralTimeseries) > 0 {
 		a, err := db.EphemeralAppender(ctx)
-		if err != nil {
-			// TODO: handle error caused by limit (ephemeral storage disabled), and report it via firstPartialErr instead.
+
+		switch {
```

Comment:
i think this switch statement would be easier to read in this structure:

```
if err != nil {
  if errors.Is(err, errEphemeralStorageDisabledForUser) {
    // first case
  } else {
    // second case
  }
} else {
  // default case
}
```

Reply:
I used that structure at first, but the code seemed less straightforward to follow. In some cases we return immediately, in others we don't. I still prefer `switch` version, that makes this more explicit and easier to read (opinions).

```golang
		if err != nil {
			if !errors.Is(err, errEphemeralStorageDisabledForUser) {
				rollback()
				return nil, err
			}

			// Ephemeral storage is disabled. Add all samples for ephemeral series as "failed".
			for _, ts := range req.EphemeralTimeseries {
				ephemeralStats.failedSamplesCount += len(ts.Samples)
			}

			updateFirstPartial(func() error {
				return fmt.Errorf(globalerror.EphemeralStorageNotEnabledForUser.Message(errEphemeralStorageDisabledForUser.Error()))
			})
			// No rollback, we will append persistent samples.
		} else {
			ephemeralApp = a.(extendedAppender)

			level.Debug(spanlog).Log("event", "got appender for ephemeral series", "ephemeralSeries", len(req.EphemeralTimeseries))

			minAppendTime, minAppendTimeAvailable := db.getEphemeralStorage().AppendableMinValidTime()

			err = i.pushSamplesToAppender(userID, req.EphemeralTimeseries, ephemeralApp, startAppend, &ephemeralStats, updateFirstPartial, nil, 0, minAppendTimeAvailable, minAppendTime, false, true)
			if err != nil {
				rollback()
				return nil, err
			}
		}
```